extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java

```diff
@@ -121,70 +122,86 @@ private static void enlistExplicitClasses(IndexView index, JpaEntitiesBuildItem
-                }
-                addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, jpaClassName);
+        if (annotations == null) {
+            return;
+        }
+
+        Set<DotName> unIndexedClasses = new HashSet<>();
+        for (AnnotationInstance annotation : annotations) {
+            AnnotationTarget target = annotation.target();
+            DotName jpaClassName;
+            switch (target.kind()) {
+                case FIELD:
+                    // TODO could fail if that's an array or a generic type
+                    jpaClassName = target.asField().type().name();
+                    break;
+                case METHOD:
+                    // TODO could fail if that's an array or a generic type
+                    jpaClassName = target.asMethod().returnType().name();
+                    break;
+                default:
+                    throw new IllegalStateException("[internal error] @Embedded placed on a unknown element: " + target);
             }
+            addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, jpaClassName,
+                    unIndexedClasses);
         }
+
+        throwIfUnIndexedClassesNotEmpty(unIndexedClasses);
     }
 
     private void enlistJPAModelClasses(IndexView index, JpaEntitiesBuildItem domainObjectCollector,
             Set<String> enumTypeCollector, DotName dotName) {
         Collection<AnnotationInstance> jpaAnnotations = index.getAnnotations(dotName);
-        if (jpaAnnotations != null && jpaAnnotations.size() > 0) {
-            for (AnnotationInstance annotation : jpaAnnotations) {
-                ClassInfo klass = annotation.target().asClass();
-                DotName targetDotName = klass.name();
-                // ignore non-jpa model classes that we think belong to JPA
-                if (nonJpaModelClasses.contains(targetDotName.toString())) {
-                    continue;
-                }
-                addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, targetDotName);
-                domainObjectCollector.addEntity(targetDotName.toString());
+        if (jpaAnnotations == null) {
+            return;
+        }
+
+        Set<DotName> unIndexedClasses = new HashSet<>();
+        for (AnnotationInstance annotation : jpaAnnotations) {
+            ClassInfo klass = annotation.target().asClass();
+            DotName targetDotName = klass.name();
+            // ignore non-jpa model classes that we think belong to JPA
+            if (nonJpaModelClasses.contains(targetDotName.toString())) {
+                continue;
             }
+            addClassHierarchyToReflectiveList(index, domainObjectCollector, enumTypeCollector, targetDotName,
+                    unIndexedClasses);
+            domainObjectCollector.addEntity(targetDotName.toString());
         }
+
+        throwIfUnIndexedClassesNotEmpty(unIndexedClasses);
```

Comment:
I usually avoid doing that as you completely lose the flow.,

I would do:
```
if (!unIndexedClasses.isEmpty()) {
    throw unindexedClassesException(unindexedClasses.isEmpty);
}
```

Reply:
Done. Moved into `discoverModelAndRegisterForReflection` with the proposed change.